source/extensions/filters/network/mysql_proxy/mysql_codec_clogin_resp.cc

```diff
@@ -15,126 +16,68 @@ namespace Extensions {
-    ClientLoginResponse::AuthSwitchMessage&& other) noexcept {
-  if (&other == this) {
-    return *this;
-  }
-  auth_plugin_data_ = std::move(other.auth_plugin_data_);
-  auth_plugin_name_ = std::move(other.auth_plugin_name_);
-  return *this;
-}
-
 bool ClientLoginResponse::AuthSwitchMessage::operator==(
     const ClientLoginResponse::AuthSwitchMessage& other) const {
   return auth_plugin_data_ == other.auth_plugin_data_ &&
          auth_plugin_name_ == other.auth_plugin_name_;
 }
 
-ClientLoginResponse::AuthMoreMessage::AuthMoreMessage(
-    ClientLoginResponse::AuthMoreMessage&& other) noexcept
-    : more_plugin_data_(std::move(other.more_plugin_data_)) {}
-
-ClientLoginResponse::AuthMoreMessage& ClientLoginResponse::AuthMoreMessage::operator=(
-    ClientLoginResponse::AuthMoreMessage&& other) noexcept {
-  if (&other == this) {
-    return *this;
-  }
-  more_plugin_data_ = std::move(other.more_plugin_data_);
-  return *this;
-}
-
 bool ClientLoginResponse::AuthMoreMessage::operator==(
     const ClientLoginResponse::AuthMoreMessage& other) const {
   return more_plugin_data_ == other.more_plugin_data_;
 }
 
-ClientLoginResponse::OkMessage::OkMessage(ClientLoginResponse::OkMessage&& other) noexcept
-    : affected_rows_(other.affected_rows_), last_insert_id_(other.last_insert_id_),
-      status_(other.status_), warnings_(other.warnings_), info_(std::move(other.info_)) {}
-
-ClientLoginResponse::OkMessage&
-ClientLoginResponse::OkMessage::operator=(ClientLoginResponse::OkMessage&& other) noexcept {
-  if (&other == this) {
-    return *this;
-  }
-  affected_rows_ = other.affected_rows_;
-  last_insert_id_ = other.last_insert_id_;
-  status_ = other.status_;
-  warnings_ = other.warnings_;
-  info_ = std::move(other.info_);
-  return *this;
-}
-
 bool ClientLoginResponse::OkMessage::operator==(const ClientLoginResponse::OkMessage& other) const {
   return affected_rows_ == other.affected_rows_ && last_insert_id_ == other.last_insert_id_ &&
          status_ == other.status_ && warnings_ == other.warnings_ && info_ == other.info_;
 }
 
-ClientLoginResponse::ErrMessage::ErrMessage(ClientLoginResponse::ErrMessage&& other) noexcept
-    : marker_(other.marker_), error_code_(other.error_code_),
-      sql_state_(std::move(other.sql_state_)), error_message_(std::move(other.error_message_)) {}
-
-ClientLoginResponse::ErrMessage&
-ClientLoginResponse::ErrMessage::operator=(ErrMessage&& other) noexcept {
-  if (&other == this) {
-    return *this;
-  }
-  error_code_ = other.error_code_;
-  marker_ = other.marker_;
-  sql_state_ = std::move(other.sql_state_);
-  error_message_ = std::move(other.error_message_);
-  return *this;
-}
-
 bool ClientLoginResponse::ErrMessage::operator==(const ErrMessage& other) const {
   return error_code_ == other.error_code_ && marker_ == other.marker_ &&
          sql_state_ == other.sql_state_ && error_message_ == other.error_message_;
 }
 
 void ClientLoginResponse::cleanup() {
```

Comment:
Actually I think it would be much cleaner and nicer NOT to use union. You have 4 message types. You could define base class called AuthMsg or Message and define some virtual methods which each message derived from base class must define. See https://github.com/envoyproxy/envoy/blob/release/v1.17/source/extensions/filters/network/postgres_proxy/postgres_message.h#L207. 
In that case your ClientLoginResponse would have to only store a single pointer to Message class. When you switch types, you could use move operator which would release previous pointer. By doing it, you pass the memory management to compiler and could delete _cleanup_ method.

Reply:
Elegant approach，